lib/range.js

```diff
@@ -0,0 +1,32 @@
+/**
+ * ...something resembling a binary search, to find the lowest line within the range.
+ * And then you could break as soon as the line is longer than the range...
+ */
+module.exports.sliceRange = (lines, startCol, endCol) => {
```

Comment:
Just wanted to write out a suggestion in pseudo code, I think you could do this:

# Phase 1.

1. Check lines[lines.length / 2]
2. If `line.startCol` > `endCol`, look at bottom half of remaining lines. If `line.startCol < endCol` look at top half of remaining lines.
3. Repeat step `2`, until you find the first index where `line.startCol > endCol`, store this as `upperIndex`.

# Phase 2.

```js
const filteredLines = []
for (let i = upperIndex - 1; i >= 0; i--) {
  const line = lines[i]
  if (startCol <= line.endCol && endCol >= line.startCol) {
    filteredLines.unshift(line)
  } else if (line.startCol < startCol) {
    break;
  }
}
```

I think picking one of either the `startCol` or `endCol` for the search will work best, f I'm thinking of the problem properly, and it seems like `startCol` is the right choice.

Reply:
@glromeo I'm not sure my mental model is right, so take what I say with a grain of salt ... but, I _think_ once we've created the array of `lines`, for line 0 - N, I believe the start/end position should never overlap, and `line[n].endPos` should always be < `line[n + 1].startPos`.

Given this, I was thinking about the algorithm, a better approach than my first recommendation might be:

* perform a binary search (starting at the half point, and going to the upper or lower half), until you find the index of the first `line` with `endCol` >= `line.startCol` -- store this as `indexStart`.

Then you can just do this:

```
const lines = []
for (let i = indexStart; i <= indexEnd; i++) {
  const line = lines[i]
  if (startCol < line.endCol && endCol >= line.startCol) {
    lines.push(line);
  } else {
    break;
  }
}
```

☝️ I think this should be functionally equivalent to the old approach, but perform way less comparisons.

Reply:
@glromeo I tested this optimization which tests half the proposed algorithm:

```js
    const lines = []
    for (const line of this.lines) {
      if (startCol <= line.endCol && endCol >= line.startCol) {
        lines.push(line)
      } else if (line.startCol > endCol) {
        break
      }
    }
```

And it works like a charm in `c8`'s test suite, the other half of the algorithm is starting at the appropriate offset of lines, rather than at `0`.